File: trading-bot/internal/strategy/scalper/scalper.go
// Package scalper implements a micro-profit scalping strategy for SGX equities.
//
// Entry logic:
//   - Spread must be at least MinSpreadTicks wide (we're paying it, so we need room)
//   - Momentum must align with intended direction (buy on up-momentum, sell on down)
//   - Volume must exceed MinVolume threshold
//
// Exit logic (first of):
//   - Price reaches TargetProfitTicks away from entry
//   - Price moves StopLossTicks against entry
//   - MaxHoldSeconds elapsed (time-based stop)
package scalper

import (
	"context"
	"fmt"
	"time"

	"github.com/mmedabo/trading-bot/config"
	"github.com/mmedabo/trading-bot/internal/broker"
	"github.com/mmedabo/trading-bot/internal/market"
	"github.com/mmedabo/trading-bot/internal/risk"
	"github.com/shopspring/decimal"
	"go.uber.org/zap"
)

// Trade tracks an open scalp position.
type Trade struct {
	Symbol      string
	OrderID     string
	Side        broker.Side
	EntryPrice  decimal.Decimal
	Qty         int64
	EnteredAt   time.Time
	TargetPrice decimal.Decimal
	StopPrice   decimal.Decimal
}

type Scalper struct {
	cfg     *config.Config
	log     *zap.Logger
	risk    *risk.Manager
	brok    broker.Broker
	history map[string]*market.TickHistory
	open    map[string]*Trade // symbol → open trade
}

func New(cfg *config.Config, log *zap.Logger, riskMgr *risk.Manager, b broker.Broker) *Scalper {
	h := make(map[string]*market.TickHistory, len(cfg.Watchlist))
	for _, sym := range cfg.Watchlist {
		h[sym] = market.NewTickHistory(cfg.Strategy.MomentumPeriod + 2)
	}
	return &Scalper{
		cfg:     cfg,
		log:     log,
		risk:    riskMgr,
		brok:    b,
		history: h,
		open:    make(map[string]*Trade),
	}
}

// OnTick is called on each new price update for a symbol.
// It updates the history and decides whether to enter or manage an open trade.
func (s *Scalper) OnTick(ctx context.Context, tick market.Tick) error {
	hist := s.history[tick.Symbol]
	if hist == nil {
		return nil
	}
	hist.Add(tick)

	if trade, ok := s.open[tick.Symbol]; ok {
		return s.manageTrade(ctx, trade, tick)
	}

	if hist.Len() < s.cfg.Strategy.MomentumPeriod {
		return nil // not enough history yet
	}

	return s.tryEnter(ctx, tick, hist)
}

func (s *Scalper) tryEnter(ctx context.Context, tick market.Tick, hist *market.TickHistory) error {
	tickSize := sgxTickSize(tick.Last)
	minSpread := tickSize.Mul(decimal.NewFromInt(int64(s.cfg.Strategy.MinSpreadTicks)))

	if tick.Spread().LessThan(minSpread) {
		return nil // spread too tight to capture profit after slippage
	}
	if tick.Volume < s.cfg.Strategy.MinVolume {
		return nil // insufficient liquidity
	}

	momentum := hist.Momentum()
	if momentum.IsZero() {
		return nil // no clear direction
	}

	var side broker.Side
	var entryPrice decimal.Decimal
	var targetPrice decimal.Decimal
	var stopPrice decimal.Decimal

	target := tickSize.Mul(decimal.NewFromInt(int64(s.cfg.Strategy.TargetProfitTicks)))
	stop := tickSize.Mul(decimal.NewFromInt(int64(s.cfg.Strategy.StopLossTicks)))

	if momentum.GreaterThan(decimal.Zero) {
		// bullish momentum: buy at ask, sell at ask + target
		side = broker.Buy
		entryPrice = tick.Ask
		targetPrice = entryPrice.Add(target)
		stopPrice = entryPrice.Sub(stop)
	} else {
		// bearish momentum: sell at bid, cover at bid - target
		// Note: short selling on SGX requires a securities borrowing arrangement;
		// in paper mode this is simulated freely.
		side = broker.Sell
		entryPrice = tick.Bid
		targetPrice = entryPrice.Sub(target)
		stopPrice = entryPrice.Add(stop)
	}

	// Position sizing: use up to MaxPositionSGD, round down to board lot (100 shares).
	// We deliberately use the full cap (not a fraction) because SGX minimum commission
	// is S$10/trade — small positions make round-trip costs uneconomical.
	qty := decimal.NewFromFloat(s.cfg.Risk.MaxPositionSGD).Div(entryPrice).IntPart()
	qty = (qty / 100) * 100 // floor to nearest board lot
	if qty < 100 {
		return nil // position too small for board lot at current price
	}

	tradeValue := entryPrice.Mul(decimal.NewFromInt(qty))
	if err := s.risk.CanOpen(tradeValue); err != nil {
		s.log.Debug("risk check failed", zap.String("symbol", tick.Symbol), zap.Error(err))
		return nil
	}

	if _, err := s.risk.ExpectedPnL(entryPrice, targetPrice, qty); err != nil {
		s.log.Debug("pnl check failed", zap.String("symbol", tick.Symbol), zap.Error(err))
		return nil
	}

	ord, err := s.brok.PlaceOrder(ctx, tick.Symbol, side, qty, entryPrice, broker.Limit)
	if err != nil {
		return fmt.Errorf("place order %s: %w", tick.Symbol, err)
	}

	s.risk.RecordOpen()
	s.open[tick.Symbol] = &Trade{
		Symbol:      tick.Symbol,
		OrderID:     ord.ID,
		Side:        side,
		EntryPrice:  entryPrice,
		Qty:         qty,
		EnteredAt:   time.Now(),
		TargetPrice: targetPrice,
		StopPrice:   stopPrice,
	}

	s.log.Info("entered trade",
		zap.String("symbol", tick.Symbol),
		zap.String("side", string(side)),
		zap.String("entry", entryPrice.String()),
		zap.String("target", targetPrice.String()),
		zap.String("stop", stopPrice.String()),
		zap.Int64("qty", qty),
	)
	return nil
}

func (s *Scalper) manageTrade(ctx context.Context, trade *Trade, tick market.Tick) error {
	heldFor := time.Since(trade.EnteredAt)
	maxHold := time.Duration(s.cfg.Strategy.MaxHoldSeconds) * time.Second

	var exitReason string
	var exitPrice decimal.Decimal

	switch trade.Side {
	case broker.Buy:
		if tick.Bid.GreaterThanOrEqual(trade.TargetPrice) {
			exitReason = "target"
			exitPrice = tick.Bid
		} else if tick.Bid.LessThanOrEqual(trade.StopPrice) {
			exitReason = "stop"
			exitPrice = tick.Bid
		}
	case broker.Sell:
		if tick.Ask.LessThanOrEqual(trade.TargetPrice) {
			exitReason = "target"
			exitPrice = tick.Ask
		} else if tick.Ask.GreaterThanOrEqual(trade.StopPrice) {
			exitReason = "stop"
			exitPrice = tick.Ask
		}
	}

	if exitReason == "" && heldFor >= maxHold {
		exitReason = "timeout"
		if trade.Side == broker.Buy {
			exitPrice = tick.Bid
		} else {
			exitPrice = tick.Ask
		}
	}

	if exitReason == "" {
		return nil // still holding
	}

	exitSide := broker.Sell
	if trade.Side == broker.Sell {
		exitSide = broker.Buy
	}

	ord, err := s.brok.PlaceOrder(ctx, trade.Symbol, exitSide, trade.Qty, exitPrice, broker.Limit)
	if err != nil {
		return fmt.Errorf("exit order %s: %w", trade.Symbol, err)
	}

	var pnl decimal.Decimal
	if trade.Side == broker.Buy {
		pnl = exitPrice.Sub(trade.EntryPrice).Mul(decimal.NewFromInt(trade.Qty))
	} else {
		pnl = trade.EntryPrice.Sub(exitPrice).Mul(decimal.NewFromInt(trade.Qty))
	}

	s.risk.RecordClose(trade.Symbol, pnl)
	delete(s.open, trade.Symbol)

	s.log.Info("exited trade",
		zap.String("symbol", trade.Symbol),
		zap.String("reason", exitReason),
		zap.String("entry", trade.EntryPrice.String()),
		zap.String("exit", exitPrice.String()),
		zap.String("pnl", pnl.String()),
		zap.String("order_id", ord.ID),
		zap.Duration("held", heldFor),
	)
	return nil
}

// OpenTrades returns a snapshot of currently open trade symbols.
func (s *Scalper) OpenTrades() []string {
	syms := make([]string, 0, len(s.open))
	for sym := range s.open {
		syms = append(syms, sym)
	}
	return syms
}

// sgxTickSize returns the minimum price movement for an SGX stock.
func sgxTickSize(price decimal.Decimal) decimal.Decimal {
	if price.LessThan(decimal.NewFromFloat(0.20)) {
		return decimal.NewFromFloat(0.001)
	}
	if price.LessThan(decimal.NewFromFloat(2.00)) {
		return decimal.NewFromFloat(0.005)
	}
	return decimal.NewFromFloat(0.01)
}

File: trading-bot-moomoo/strategy/scalper.py
"""
SGX Micro-Profit Scalping Strategy
====================================
Entry:  positive momentum (5-tick window) + spread ≥ 2 ticks + trade is profitable after commission
Exit:   first of — target ticks hit | stop-loss ticks hit | max hold time elapsed
"""
from __future__ import annotations
import logging
import time as _time
from collections import deque
from dataclasses import dataclass, field
from datetime import datetime

from risk.manager import RiskManager

log = logging.getLogger("scalper")


# SGX minimum tick sizes by price band
def tick_size(price: float) -> float:
    if price < 0.20:
        return 0.001
    if price < 2.00:
        return 0.005
    return 0.01


@dataclass
class Quote:
    symbol: str
    bid: float
    ask: float
    last: float
    volume: int
    ts: float = field(default_factory=_time.time)

    @property
    def spread(self) -> float:
        return self.ask - self.bid

    @property
    def mid(self) -> float:
        return (self.bid + self.ask) / 2


@dataclass
class OpenTrade:
    symbol: str
    side: str           # "BUY" or "SELL"
    entry_price: float
    qty: int
    target_price: float
    stop_price: float
    order_id: str
    entered_at: float = field(default_factory=_time.time)


class Scalper:
    def __init__(self, cfg: dict, risk: RiskManager, broker):
        self._cfg = cfg
        self._risk = risk
        self._broker = broker
        self._history: dict[str, deque[Quote]] = {}
        self._open: dict[str, OpenTrade] = {}   # symbol → trade
        window = cfg["momentum_period"] + 2
        for sym in cfg.get("watchlist", []):
            self._history[sym] = deque(maxlen=window)

    # ------------------------------------------------------------------ #

    def on_quote(self, q: Quote):
        """Called each polling cycle for a symbol."""
        hist = self._history.setdefault(q.symbol, deque(maxlen=self._cfg["momentum_period"] + 2))
        hist.append(q)

        if q.symbol in self._open:
            self._manage(q)
        elif len(hist) >= self._cfg["momentum_period"]:
            self._try_enter(q, hist)

    # ------------------------------------------------------------------ #

    def _try_enter(self, q: Quote, hist: deque[Quote]):
        tick = tick_size(q.last)
        min_spread = tick * self._cfg["min_spread_ticks"]

        if q.spread < min_spread:
            return
        if q.volume < self._cfg.get("min_volume", 0):
            return

        momentum = hist[-1].mid - hist[0].mid
        if momentum == 0:
            return

        target_ticks = self._cfg["target_profit_ticks"]
        stop_ticks   = self._cfg["stop_loss_ticks"]

        if momentum > 0:
            side         = "BUY"
            entry        = q.ask
            target_price = entry + tick * target_ticks
            stop_price   = entry - tick * stop_ticks
        else:
            side         = "SELL"
            entry        = q.bid
            target_price = entry - tick * target_ticks
            stop_price   = entry + tick * stop_ticks

        # Position sizing: as many board lots as MaxPositionSGD allows
        max_sgd = self._cfg["max_position_sgd"]
        qty = int((max_sgd / entry) // 100) * 100   # floor to board lot
        if qty < 100:
            log.debug("%s: price too high for board lot at S$%s limit", q.symbol, max_sgd)
            return

        trade_val = entry * qty
        ok, reason = self._risk.can_open(trade_val)
        if not ok:
            log.debug("%s: risk blocked — %s", q.symbol, reason)
            return

        net = self._risk.expected_net_pnl(entry, target_price, qty)
        if net <= 0:
            log.debug("%s: not profitable after commission (net=%.3f)", q.symbol, net)
            return

        order_id = self._broker.place_order(
            symbol=q.symbol, side=side, qty=qty, price=entry,
        )
        if order_id is None:
            log.warning("%s: order placement failed", q.symbol)
            return

        self._risk.record_open()
        self._open[q.symbol] = OpenTrade(
            symbol=q.symbol, side=side, entry_price=entry, qty=qty,
            target_price=target_price, stop_price=stop_price, order_id=order_id,
        )
        log.info(
            "ENTER %s %s | entry=%.4f target=%.4f stop=%.4f qty=%d net_exp=%.2f SGD",
            side, q.symbol, entry, target_price, stop_price, qty, net,
        )

    def _manage(self, q: Quote):
        trade = self._open[q.symbol]
        max_hold = self._cfg["max_hold_seconds"]
        held = _time.time() - trade.entered_at

        exit_reason = None
        exit_price  = None

        if trade.side == "BUY":
            if q.bid >= trade.target_price:
                exit_reason, exit_price = "target", q.bid
            elif q.bid <= trade.stop_price:
                exit_reason, exit_price = "stop",   q.bid
        else:  # SELL
            if q.ask <= trade.target_price:
                exit_reason, exit_price = "target", q.ask
            elif q.ask >= trade.stop_price:
                exit_reason, exit_price = "stop",   q.ask

        if exit_reason is None and held >= max_hold:
            exit_reason = "timeout"
            exit_price  = q.bid if trade.side == "BUY" else q.ask

        if exit_reason is None:
            return

        exit_side = "SELL" if trade.side == "BUY" else "BUY"
        self._broker.place_order(
            symbol=trade.symbol, side=exit_side, qty=trade.qty, price=exit_price,
        )

        if trade.side == "BUY":
            pnl = (exit_price - trade.entry_price) * trade.qty
        else:
            pnl = (trade.entry_price - exit_price) * trade.qty

        self._risk.record_close(trade.symbol, pnl)
        del self._open[trade.symbol]

        log.info(
            "EXIT  %s | reason=%-8s entry=%.4f exit=%.4f pnl=%+.2f SGD held=%.1fs",
            trade.symbol, exit_reason, trade.entry_price, exit_price, pnl, held,
        )

    @property
    def open_symbols(self) -> list[str]:
        return list(self._open.keys())
